### PR TITLE
Add Package matching function for finding packages with fields

### PR DESCRIPTION
**Pull request by lseppala**
This change provides a new method `matching` on `Package` objects. This method takes a simple "matcher" object with optional keys `namespace`, `name`, and `version`, and returns true if all specified keys match values of the Package.

Similarly, it also provides a `packagesMatching` method on `PackageCache`. Given a matcher, it returns all stored `Package` objects matching the matcher.

### CODE REVIEW

File: src/snapshot.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ Core functionality for creating a snapshot of a project's dependencies.
 
 /**
  * When multiple snapshots are submit, Job provides the means for Snapshots to
```
Comment (juxtin) [inline src/snapshot.ts:14]: I know this isn't part of this diff, but
```suggestion
 * When multiple snapshots are submitted, Job provides the means for Snapshots to
```

**Review by juxtin — APPROVED** [review]
Looks like a nice addition! I just had one tiny unrelated typo I had to point out.

File: example/action-dist/index.js [hunk 2]
```diff
@@ -259,8 +270,7 @@ const packageurl_js_1 = __nccwpck_require2_(8915);
  * We consider all packages that are defined in the [Package URL spec](https://github.com/package-url/purl-spec/blob/1eae1e95d81fddf8ae7f06b4dfc7b5b5be0cc3e2/PURL-TYPES.rst) as being valid package types.
  */
 class Package {
-    /**
-     * A Package can be constructed with a PackageURL or a string conforming to
+    /** A Package can be constructed with a PackageURL or a string conforming to
```
Comment (pcarlisle) [inline example/action-dist/index.js:273]: is this change intended?

File: src/package.ts [hunk 3]
```diff
@@ -15,8 +15,7 @@ export class Package {
    */
   dependencies: Array<Package> // eslint-disable-line no-use-before-define
 
-  /**
-   * A Package can be constructed with a PackageURL or a string conforming to
+  /** A Package can be constructed with a PackageURL or a string conforming to
```
Comment (pcarlisle) [inline src/package.ts:18]: i guess this is the source of that and I commented on a generated file

File: src/package-cache.test.ts [hunk 4]
```diff
@@ -39,5 +39,47 @@ describe('PackageCache', () => {
       // purposely using reference equality with 'toBe'
       expect(cache.package(purl)).toBe(dep)
     })
+    it('.packagesMatching returns package that match matcher', () => {
```
Comment (pcarlisle) [inline src/package-cache.test.ts:42]:
```suggestion
    it('.packagesMatching returns package that matches matcher', () => {
```

File: src/package.ts [hunk 5]
```diff
@@ -85,4 +93,26 @@ export class Package {
...
+   */
+  matching(matcher: {
+    namespace?: string
+    name?: string
+    version?: string
+  }): boolean {
+    return (
+      (matcher.namespace === undefined ||
+        this.packageURL.namespace === matcher.namespace) &&
+      (matcher.name === undefined || this.packageURL.name === matcher.name) &&
```
Comment (pcarlisle) [inline src/package.ts:113]: this would be slightly more readable if you indented each clause the same way

**Review by pcarlisle — APPROVED** [review]
my comments are all aesthetic and you can take them or leave them

File: src/package.ts (hunk 5 above)
Comment (lseppala) [reply src/package.ts:113]: I agree. I was fighting with the auto-formatter (`prettier`, invoked by `npm run format` and `npm run all`). The auto-formatter won 😥

File: example/action-dist/index.js (hunk 2 above)
Comment (lseppala) [reply example/action-dist/index.js]: Fixed in source file

File: src/package.ts (hunk 5 above)
Comment (lseppala) [reply src/package.ts]: fixed with an ignore line
```
// prettier ignore
```
